Mathlib/Order/OrderType.lean

```diff
@@ -0,0 +1,450 @@
+  cases nonempty_unique α
+  exact (@ofUniqueOfIrrefl _).ordertype_congr
+
+@[simp]
+theorem type_eq_one_iff_unique [LinearOrder α] : type α = 1 ↔ Nonempty (Unique α) :=
+  ⟨fun h ↦ let ⟨s⟩ := type_eq.1 h; ⟨s.toEquiv.unique⟩,
+    fun ⟨_⟩ ↦ type_eq_one_of_unique⟩
+
+theorem type_pUnit : type PUnit = 1 :=
+  rfl
+
+theorem type_unit : type Unit = 1 :=
+  rfl
+
+@[simp]
+theorem toType_empty_iff_eq_zero {o : OrderType} : IsEmpty o.toType ↔ o = 0 := by
+  rw [← @type_eq_zero_iff_isEmpty o.toType, type_ordtoType]
+
+instance isEmpty_toType_zero : IsEmpty (toType 0) :=
+  toType_empty_iff_eq_zero.2 rfl
+
+@[simp]
+theorem toType_nonempty_iff_ne_zero {o : OrderType} : Nonempty o.toType ↔ o ≠ 0 := by
+  rw [← @type_ne_zero_iff_nonempty o.toType, type_ordtoType]
+
+protected theorem one_ne_zero : (1 : OrderType) ≠ 0 :=
+  type_ne_zero_of_nonempty
+
+instance nontrivial : Nontrivial OrderType.{u} :=
+  ⟨⟨1, 0, OrderType.one_ne_zero⟩⟩
+
+/-- `Quotient.inductionOn` specialized to OrderTypes. -/
+@[elab_as_elim]
+theorem inductionOn {C : OrderType → Prop} (o : OrderType)
+    (H : ∀ α [LinearOrder α], C (type α)) : C o :=
+  Quot.inductionOn o (fun α ↦ H α)
+
+/-- `Quotient.inductionOn₂` specialized to OrderTypes. -/
+@[elab_as_elim]
+theorem inductionOn₂ {C : OrderType → OrderType → Prop} (o₁ o₂ : OrderType)
+    (H : ∀ α [LinearOrder α] β [LinearOrder β], C (type α) (type β)) : C o₁ o₂ :=
+  Quotient.inductionOn₂ o₁ o₂ fun α β ↦ H α β
+
+/-- `Quotient.inductionOn₃` specialized to OrderTypes. -/
+@[elab_as_elim]
+theorem inductionOn₃ {C : OrderType → OrderType → OrderType → Prop} (o₁ o₂ o₃ : OrderType)
+    (H : ∀ α [LinearOrder α] β [LinearOrder β] γ [LinearOrder γ],
+      C (type α) (type β) (type γ)) : C o₁ o₂ o₃ :=
+  Quotient.inductionOn₃ o₁ o₂ o₃ fun α β γ ↦
+    H α β γ
+
+open Classical in
+/-- To prove a result on OrderTypes, it suffices to prove it for order types of linear orders. -/
+@[elab_as_elim]
+theorem inductionOnLinOrd {C : OrderType → Prop} (o : OrderType)
+    (H : ∀ α [LinearOrder α], C (type α)) : C o :=
+  inductionOn o fun α ↦ H α
+
+open Classical in
+/-- To define a function on OrderTypes, it suffices to define them on all linear order isomorphisms.
+-/
+def liftOnLinOrd {δ : Sort v} (o : OrderType) (f : ∀ (α) [LinearOrder α], δ)
+    (c : ∀ (α) [LinearOrder α] (β) [LinearOrder β],
+      type α = type β → f α = f β) : δ :=
+  Quotient.liftOn o (fun w ↦ f w)
+    fun w₁ w₂ h ↦ c w₁ w₂ (Quotient.sound h)
+
+@[simp]
+theorem liftOnLinOrd_type {δ : Sort v} (f : ∀ (α) [LinearOrder α], δ)
+    (c : ∀ (α) [LinearOrder α] (β) [LinearOrder β],
+      type α = type β → f α = f β) {γ} [LinearOrder γ] :
+    liftOnLinOrd (type γ) f c = f γ := by
+  change Quotient.liftOn' ⟦_⟧ _ _ = _
+  rw [Quotient.liftOn'_mk]
+
+/-! ### The order on OrderTypes -/
+
+/--
+For `OrderType`:
+
+Less-than-or-equal is defined such that linear orders `r` on `α` and `s` on `β`
+satisfy `type α ≤ type β` if there exists an order embedding from `α` to `β`.
```

Comment:
I think we can make this a bit less verbose:
```suggestion
The order is defined so that `type α ≤ type β` iff there exists an order embedding `α ↪o β`.
```